cuML/src/glm/qn/simple_mat.h

```diff
@@ -294,55 +294,48 @@ std::ostream &operator<<(std::ostream &os, const SimpleMat<T> &mat) {
   return os;
 }
 
-// TODO The following classes own their storage and are just for testing
-// The internals should be replaced by device_buffer+device_allocator, when ready
 template <typename T> struct SimpleVecOwning : SimpleVec<T> {
   typedef SimpleVec<T> Super;
-  bool allocated;
+  typedef MLCommon::device_buffer<T> Buffer;
+  std::unique_ptr<Buffer> buf;
 
-  SimpleVecOwning() : Super(), allocated(false) {}
+  SimpleVecOwning() : Super() {}
 
-  SimpleVecOwning(int n) : Super(), allocated(false) { reset(n); }
-
-  ~SimpleVecOwning() noexcept(false) { CUDA_CHECK(cudaFree(Super::data)); }
+  SimpleVecOwning(const cumlHandle_impl &handle, int n, cudaStream_t stream)
+      : Super() {
+    reset(handle, n, stream);
+  }
 
-  void reset(int n) {
-    if (allocated) {
-      CUDA_CHECK(cudaFree(this->data));
-    }
-    MLCommon::allocate(this->data, n);
-    Super::reset(this->data, n);
-    allocated = true;
+  void reset(const cumlHandle_impl &handle, int n, cudaStream_t stream) {
+    buf.reset(new Buffer(handle.getDeviceAllocator(), stream, n));
```

Comment:
Could you simply call `buf->resize(n,stream)` instead?

Reply:
Do you have a strong preference for this?
This would mean that the default constructor of `SimpleMatOwning` need the allocator as well to create an empty buffer object. As it is, it might be clearer, when the actual allocation happens.

Reply:
As it probably makes sense to rework the different tensor/matrix abstractions to be generally usable across different cuML algorithms some time I do not have a strong preference here. Just to explain my thoughts: I think it makes sense that a class that owns memory (and should use the provided allocator infrastructure for that) needs an allocator object to get constructed. So my preference would be do delete the default constructor.  Does that make sense to you?

Reply:
Yeah, makes a lot of sense. I removed the default constructor